From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: hayanesuru <hayanesuru@outlook.jp>
Date: Tue, 3 Jun 2025 15:20:59 +0900
Subject: [PATCH] optimize mob spawning

Avoid getChunk calls if its position is same as the chunk used for mob spawning

Cache chunk generator

Don't sync load chunks for mob spawning checks

Fix data race in async mob spawning
by adding chunk position to the mob count map
then apply result on server thread.

Generally faster than the non-async approach

iterate over all entities, get their chunk, and increment the count

diff --git a/net/minecraft/server/level/ChunkMap.java b/net/minecraft/server/level/ChunkMap.java
index 5c369b3d94e369c3f240821ad90b9d96223f24ca..9803c395fce103cb7bc746f43a017ff9ed99728c 100644
--- a/net/minecraft/server/level/ChunkMap.java
+++ b/net/minecraft/server/level/ChunkMap.java
@@ -278,6 +278,7 @@ public class ChunkMap extends ChunkStorage implements ChunkHolder.PlayerProvider
     }
     // Paper end - per player mob count backoff
     public int getMobCountNear(final ServerPlayer player, final net.minecraft.world.entity.MobCategory mobCategory) {
+        // Leaf - diff - async mob spawning - optimize mob spawning
         return player.mobCounts[mobCategory.ordinal()] + player.mobBackoffCounts[mobCategory.ordinal()]; // Paper - per player mob count backoff
      }
     // Paper end - Optional per player mob spawns
diff --git a/net/minecraft/server/level/ServerChunkCache.java b/net/minecraft/server/level/ServerChunkCache.java
index 3a6c894178829cec8daa08ea9f0294f7f39a8efe..ed267cb678206e3ced43e1aafbac203ff87bee74 100644
--- a/net/minecraft/server/level/ServerChunkCache.java
+++ b/net/minecraft/server/level/ServerChunkCache.java
@@ -70,11 +70,11 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
     private final long[] lastChunkPos = new long[4];
     private final ChunkStatus[] lastChunkStatus = new ChunkStatus[4];
     private final ChunkAccess[] lastChunk = new ChunkAccess[4];
-    private final List<LevelChunk> spawningChunks = new ObjectArrayList<>();
+    private final List<LevelChunk> spawningChunks = it.unimi.dsi.fastutil.objects.ReferenceArrayList.wrap(new LevelChunk[0]); // Leaf - optimize mob spawning
     private final Set<ChunkHolder> chunkHoldersToBroadcast = new ReferenceOpenHashSet<>();
     @Nullable
     @VisibleForDebug
-    private NaturalSpawner.SpawnState lastSpawnState;
+    private volatile NaturalSpawner.SpawnState lastSpawnState; // Leaf - optimize mob spawning
     // Paper start
     public final ca.spottedleaf.concurrentutil.map.ConcurrentLong2ReferenceChainedHashTable<net.minecraft.world.level.chunk.LevelChunk> fullChunks = new ca.spottedleaf.concurrentutil.map.ConcurrentLong2ReferenceChainedHashTable<>();
     public int getFullChunksCount() {
@@ -183,8 +183,8 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
     }
     // Paper end - chunk tick iteration optimisations
 
-    public boolean firstRunSpawnCounts = true; // Pufferfish
-    public final java.util.concurrent.atomic.AtomicBoolean _pufferfish_spawnCountsReady = new java.util.concurrent.atomic.AtomicBoolean(false); // Pufferfish - optimize countmobs
+    public final java.util.concurrent.atomic.AtomicInteger _pufferfish_spawnCountsReady = new java.util.concurrent.atomic.AtomicInteger(); // Leaf - optimize mob spawning
+    public static final MobCategory[] EMPTY_MOB_CATEGORIES = new MobCategory[0]; // Leaf - optimize mob spawning
 
     public ServerChunkCache(
         ServerLevel level,
@@ -510,54 +510,6 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
 
             this.broadcastChangedChunks(); // Gale - Purpur - remove vanilla profiler
         }
-
-        // Pufferfish start - optimize mob spawning
-        if (org.dreeam.leaf.config.modules.async.AsyncMobSpawning.enabled) {
-            for (ServerPlayer player : this.level.players) {
-                // Paper start - per player mob spawning backoff
-                for (int ii = 0; ii < ServerPlayer.MOBCATEGORY_TOTAL_ENUMS; ii++) {
-                    player.mobCounts[ii] = 0;
-
-                    int newBackoff = player.mobBackoffCounts[ii] - 1; // TODO make configurable bleed // TODO use nonlinear algorithm?
-                    if (newBackoff < 0) {
-                        newBackoff = 0;
-                    }
-                    player.mobBackoffCounts[ii] = newBackoff;
-                }
-                // Paper end - per player mob spawning backoff
-            }
-            if (firstRunSpawnCounts) {
-                firstRunSpawnCounts = false;
-                _pufferfish_spawnCountsReady.set(true);
-            }
-            if (_pufferfish_spawnCountsReady.getAndSet(false)) {
-                net.minecraft.server.MinecraftServer.getServer().mobSpawnExecutor.submit(() -> {
-                    int mapped = distanceManager.getNaturalSpawnChunkCount();
-                    ca.spottedleaf.moonrise.common.list.IteratorSafeOrderedReferenceSet.Iterator<Entity> objectiterator =
-                        level.entityTickList.entities.iterator(ca.spottedleaf.moonrise.common.list.IteratorSafeOrderedReferenceSet.ITERATOR_FLAG_SEE_ADDITIONS);
-                    try {
-                        gg.pufferfish.pufferfish.util.IterableWrapper<Entity> wrappedIterator =
-                            new gg.pufferfish.pufferfish.util.IterableWrapper<>(objectiterator);
-                        // Fix: Use proper mob cap calculator based on configuration
-                        LocalMobCapCalculator mobCapCalculator = !level.paperConfig().entities.spawning.perPlayerMobSpawns ?
-                            new LocalMobCapCalculator(chunkMap) : null;
-
-                        // This ensures the caps are properly enforced by using the correct calculator
-                        lastSpawnState = NaturalSpawner.createState(
-                            mapped,
-                            wrappedIterator,
-                            ServerChunkCache.this::getFullChunk,
-                            mobCapCalculator,  // This is the key fix - was previously null
-                            level.paperConfig().entities.spawning.perPlayerMobSpawns
-                        );
-                    } finally {
-                        objectiterator.finishedIterating();
-                    }
-                    _pufferfish_spawnCountsReady.set(true);
-                });
-            }
-        }
-        // Pufferfish end
     }
 
     private void broadcastChangedChunks() { // Gale - Purpur - remove vanilla profiler
@@ -574,9 +526,9 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
     private void tickChunks(long timeInhabited) { // Gale - Purpur - remove vanilla profiler
         int naturalSpawnChunkCount = this.distanceManager.getNaturalSpawnChunkCount();
         // Paper start - Optional per player mob spawns
-        NaturalSpawner.SpawnState spawnState;
+        //NaturalSpawner.SpawnState spawnState;
+        java.util.concurrent.atomic.AtomicInteger state = _pufferfish_spawnCountsReady;
         if ((this.spawnFriendlies || this.spawnEnemies) && this.level.paperConfig().entities.spawning.perPlayerMobSpawns) { // don't count mobs when animals and monsters are disabled
-            if (!org.dreeam.leaf.config.modules.async.AsyncMobSpawning.enabled) { // Pufferfish - moved down when async processing
             // re-set mob counts
             for (ServerPlayer player : this.level.players) {
                 // Paper start - per player mob spawning backoff
@@ -591,19 +543,21 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
                 }
                 // Paper end - per player mob spawning backoff
             }
-            lastSpawnState = NaturalSpawner.createState(naturalSpawnChunkCount, this.level.getAllEntities(), this::getFullChunk, null, true); // Pufferfish - async mob spawning
-            } // Pufferfish - (endif) moved down when async processing
+            // Leaf start - optimize mob spawning
+            if (!org.dreeam.leaf.config.modules.async.AsyncMobSpawning.enabled) {
+                lastSpawnState = NaturalSpawner.createState(naturalSpawnChunkCount, this.level.getAllEntities(), this::getFullChunk, null, true); // Pufferfish - async mob spawning
+                state.set(2);
+            }
+            // Leaf end - optimize mob spawning
         } else {
-            // Pufferfish start - async mob spawning
-            lastSpawnState = NaturalSpawner.createState(naturalSpawnChunkCount, this.level.getAllEntities(), this::getFullChunk, !this.level.paperConfig().entities.spawning.perPlayerMobSpawns ? new LocalMobCapCalculator(this.chunkMap) : null, false);
-            _pufferfish_spawnCountsReady.set(true);
-            // Pufferfish end
+            lastSpawnState = NaturalSpawner.createState(naturalSpawnChunkCount, this.level.getAllEntities(), this::getFullChunk, !this.level.paperConfig().entities.spawning.perPlayerMobSpawns ? new LocalMobCapCalculator(this.chunkMap) : null, false); // Leaf - optimize mob spawning
+            state.set(2); // Leaf - optimize mob spawning
         }
         // Paper end - Optional per player mob spawns
         //this.lastSpawnState = spawnState; // Pufferfish - this is managed asynchronously
         boolean _boolean = this.level.getGameRules().getBoolean(GameRules.RULE_DOMOBSPAWNING) && !this.level.players().isEmpty(); // CraftBukkit
         int _int = this.level.getGameRules().getInt(GameRules.RULE_RANDOMTICKING);
-        List<MobCategory> filteredSpawningCategories;
+        MobCategory[] filteredSpawningCategories; // Leaf - optimize mob spawning
         if (_boolean && (this.spawnEnemies || this.spawnFriendlies)) {
             // Paper start - PlayerNaturallySpawnCreaturesEvent
             for (ServerPlayer entityPlayer : this.level.players()) {
@@ -611,27 +565,62 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
                 chunkRange = Math.min(chunkRange, 8);
                 entityPlayer.playerNaturallySpawnedEvent = new com.destroystokyo.paper.event.entity.PlayerNaturallySpawnCreaturesEvent(entityPlayer.getBukkitEntity(), (byte) chunkRange);
                 entityPlayer.playerNaturallySpawnedEvent.callEvent();
+                this.level.natureSpawnChunkMap.addPlayer(entityPlayer); // Leaf - optimize mob spawning
             }
             // Paper end - PlayerNaturallySpawnCreaturesEvent
             boolean flag = this.level.ticksPerSpawnCategory.getLong(org.bukkit.entity.SpawnCategory.ANIMAL) != 0L && this.level.getLevelData().getGameTime() % this.level.ticksPerSpawnCategory.getLong(org.bukkit.entity.SpawnCategory.ANIMAL) == 0L; // CraftBukkit
-            filteredSpawningCategories = NaturalSpawner.getFilteredSpawningCategories(lastSpawnState, this.spawnFriendlies, this.spawnEnemies, flag, this.level); // CraftBukkit // Pufferfish
+            filteredSpawningCategories = NaturalSpawner.getFilteredSpawningCategories(lastSpawnState, this.spawnFriendlies, this.spawnEnemies, flag, this.level).toArray(EMPTY_MOB_CATEGORIES); // CraftBukkit // Pufferfish // Leaf - optimize mob spawning
         } else {
-            filteredSpawningCategories = List.of();
+            filteredSpawningCategories = EMPTY_MOB_CATEGORIES; // Leaf - optimize mob spawning
         }
 
         List<LevelChunk> list = this.spawningChunks;
 
         try {
-            this.chunkMap.collectSpawningChunks(list);
+            // Leaf start - optimize mob spawning
+            // this.chunkMap.collectSpawningChunks(list);
+            this.level.natureSpawnChunkMap.build(this.level);
+            this.level.natureSpawnChunkMap.collectSpawningChunks(this.level.moonrise$getPlayerTickingChunks(), list);
             // Paper start - chunk tick iteration optimisation
             this.shuffleRandom.setSeed(this.level.random.nextLong());
             if (!this.level.paperConfig().entities.spawning.perPlayerMobSpawns) Util.shuffle(list, this.shuffleRandom); // Paper - Optional per player mob spawns; do not need this when per-player is enabled
             // Paper end - chunk tick iteration optimisation
 
-            for (LevelChunk levelChunk : list) {
-                this.tickSpawningChunk(levelChunk, timeInhabited, filteredSpawningCategories, lastSpawnState); // Pufferfish
+            boolean enqueue = org.dreeam.leaf.config.modules.async.AsyncMobSpawning.enabled
+                && (this.spawnFriendlies || this.spawnEnemies)
+                && this.level.paperConfig().entities.spawning.perPlayerMobSpawns;
+            boolean cas = state.compareAndSet(2, 0);
+            if (!enqueue || cas) {
+                NaturalSpawner.SpawnState spawnState = cas ? lastSpawnState : null;
+                if (spawnState != null) {
+                    spawnState.applyPerPlayerMobCount(this.level);
+                }
+                MobCategory[] categories = spawnState == null ? EMPTY_MOB_CATEGORIES : filteredSpawningCategories;
+                ChunkGenerator generator = this.getGenerator();
+                if (list instanceof it.unimi.dsi.fastutil.objects.ReferenceArrayList<LevelChunk> list1) {
+                    LevelChunk[] raw = list1.elements();
+                    for (int i = 0, j = list1.size(); i < j; i++) {
+                        this.tickSpawningChunk(raw[i], timeInhabited, categories, spawnState, generator);
+                    }
+                } else {
+                    for (LevelChunk levelChunk : list) {
+                        this.tickSpawningChunk(levelChunk, timeInhabited, categories, spawnState, generator);
+                    }
+                }
             }
+            if (enqueue && state.compareAndSet(0, 1)) {
+                ServerLevel level = this.level;
+                int mapped = this.distanceManager.getNaturalSpawnChunkCount();
+                Iterable<Entity> entities = this.level.getAllEntities();
+                level.getServer().mobSpawnExecutor.submit(() -> {
+                    NaturalSpawner.SpawnState newState = NaturalSpawner.createStateAsync(mapped, entities, level);
+                    ServerChunkCache.this.lastSpawnState = state.get() != 1 ? null : newState;
+                    state.set(2);
+                });
+            }
+            // Leaf end - optimize mob spawning
         } finally {
+            this.level.natureSpawnChunkMap.clear(); // Leaf - optimize mob spawning
             list.clear();
         }
 
@@ -641,16 +630,21 @@ public class ServerChunkCache extends ChunkSource implements ca.spottedleaf.moon
         }
     }
 
+    // Leaf start - optimize mob spawning - cache ChunkGenerator
     private void tickSpawningChunk(LevelChunk chunk, long timeInhabited, List<MobCategory> spawnCategories, NaturalSpawner.SpawnState spawnState) {
+        tickSpawningChunk(chunk, timeInhabited, spawnCategories.toArray(EMPTY_MOB_CATEGORIES), spawnState, getGenerator());
+    }
+    private void tickSpawningChunk(LevelChunk chunk, long timeInhabited, MobCategory[] spawnCategories, NaturalSpawner.SpawnState spawnState, ChunkGenerator generator) {
+    // Leaf end - optimize mob spawning - cache ChunkGenerator
         ChunkPos pos = chunk.getPos();
         chunk.incrementInhabitedTime(timeInhabited);
         if (true) { // Paper - rewrite chunk system
             this.level.tickThunder(chunk);
         }
 
-        if (!spawnCategories.isEmpty()) {
-            if (this.level.getWorldBorder().isWithinBounds(pos) && (!org.dreeam.leaf.config.modules.async.AsyncMobSpawning.enabled || _pufferfish_spawnCountsReady.get())) { // Paper - rewrite chunk system // Pufferfish
-                NaturalSpawner.spawnForChunk(this.level, chunk, spawnState, spawnCategories);
+        if (spawnCategories.length != 0) {
+            if (this.level.getWorldBorder().isWithinBounds(pos)) { // Paper - rewrite chunk system
+                NaturalSpawner.spawnForChunk(this.level, chunk, spawnState, spawnCategories, generator); // Leaf - optimize mob spawning - cache ChunkGenerator
             }
         }
     }
diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index b3d877089f4eebcd78f1019a91d9d71615a15148..b9986a9df48f6986c8151d727893b750557e9a74 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1092,6 +1092,7 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
     private int currentIceAndSnowTick = 0; protected void resetIceAndSnowTick() { this.currentIceAndSnowTick = this.simpleRandom.nextInt(16); } // Gale - Airplane - optimize random calls in chunk ticking
 
     public final org.dreeam.leaf.world.DespawnMap despawnMap = new org.dreeam.leaf.world.DespawnMap(paperConfig()); // Leaf - optimize despawn
+    public final org.dreeam.leaf.world.NatureSpawnChunkMap natureSpawnChunkMap = new org.dreeam.leaf.world.NatureSpawnChunkMap(); // Leaf - optimize mob spawning
     public void tickChunk(LevelChunk chunk, int randomTickSpeed) {
         final net.minecraft.world.level.levelgen.BitRandomSource simpleRandom = this.simpleRandom; // Paper - optimise random ticking // Leaf - Faster random generator - upcasting
         ChunkPos pos = chunk.getPos();
diff --git a/net/minecraft/util/random/WeightedList.java b/net/minecraft/util/random/WeightedList.java
index daab145028c59c07049a2cf75c4f159e4c5073dc..221976c3b3a4a4d6d00bc4eb13fc1859c1910271 100644
--- a/net/minecraft/util/random/WeightedList.java
+++ b/net/minecraft/util/random/WeightedList.java
@@ -70,6 +70,17 @@ public class WeightedList<E> { // Paper - non-final
         }
     }
 
+    // Leaf start - optimize mob spawning
+    @Nullable
+    public E getRandomNullable(RandomSource random) {
+        if (this.selector == null) {
+            return null;
+        } else {
+            return this.selector.get(random.nextInt(this.totalWeight));
+        }
+    }
+    // Leaf end - optimize mob spawning
+
     public E getRandomOrThrow(RandomSource random) {
         if (this.selector == null) {
             throw new IllegalStateException("Weighted list has no elements");
diff --git a/net/minecraft/world/level/NaturalSpawner.java b/net/minecraft/world/level/NaturalSpawner.java
index bb655318f49242858e2c25d5469705c0c314ed85..c1ccb03768e54f144b8978cf27b6bc7b39bcef77 100644
--- a/net/minecraft/world/level/NaturalSpawner.java
+++ b/net/minecraft/world/level/NaturalSpawner.java
@@ -68,6 +68,7 @@ public final class NaturalSpawner {
         return createState(spawnableChunkCount, entities, chunkGetter, calculator, false);
     }
 
+    @Deprecated // Leaf - optimize mob spawning
     public static NaturalSpawner.SpawnState createState(
         int spawnableChunkCount, Iterable<Entity> entities, NaturalSpawner.ChunkGetter chunkGetter, LocalMobCapCalculator calculator, final boolean countMobs
     ) {
@@ -111,6 +112,63 @@ public final class NaturalSpawner {
         return new NaturalSpawner.SpawnState(spawnableChunkCount, map, potentialCalculator, calculator);
     }
 
+    // Leaf start - optimize mob spawning
+    public static NaturalSpawner.SpawnState createStateAsync(
+        int spawnableChunkCount, Iterable<Entity> entities, ServerLevel level
+    ) {
+        // Paper end - Optional per player mob spawns
+        PotentialCalculator potentialCalculator = new PotentialCalculator();
+        Object2IntOpenHashMap<MobCategory> map = new Object2IntOpenHashMap<>();
+        it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap<int[]> chunkCap = new it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap<>();
+        boolean countAllMobsForSpawning = level.paperConfig().entities.spawning.countAllMobsForSpawning;
+        for (Entity entity : entities) {
+            if (!(entity instanceof Mob mob && (mob.isPersistenceRequired() || mob.requiresCustomPersistence()))) {
+                MobCategory category = entity.getType().getCategory();
+                if (category != MobCategory.MISC) {
+                    // Paper start - Only count natural spawns
+                    if (!countAllMobsForSpawning &&
+                        !(entity.spawnReason == org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.NATURAL ||
+                            entity.spawnReason == org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.CHUNK_GEN)) {
+                        continue;
+                    }
+                    // Paper end - Only count natural spawns
+                    BlockPos blockPos = entity.blockPosition();
+                    LevelChunk chunk = level.chunkSource.fullChunks.get(ChunkPos.asLong(blockPos));
+                    if (chunk != null) {
+                        MobSpawnSettings.MobSpawnCost mobSpawnCost = getRoughBiome(blockPos, chunk).getMobSettings().getMobSpawnCost(entity.getType());
+                        if (mobSpawnCost != null) {
+                            potentialCalculator.addCharge(entity.blockPosition(), mobSpawnCost.charge());
+                        }
+
+                        map.addTo(category, 1);
+                        // Paper start - Optional per player mob spawns
+                        final int index = entity.getType().getCategory().ordinal();
+                        ++chunkCap.computeIfAbsent(chunk.coordinateKey, k -> new int[net.minecraft.server.level.ServerPlayer.MOBCATEGORY_TOTAL_ENUMS])[index];
+                        /*
+                        final int index = entity.getType().getCategory().ordinal();
+                        final var inRange = level.moonrise$getNearbyPlayers().getPlayers(entity.chunkPosition(), NearbyPlayers.NearbyMapType.TICK_VIEW_DISTANCE);
+                        if (inRange == null) {
+                            continue;
+                        }
+
+                        final net.minecraft.server.level.ServerPlayer[] backingSet = inRange.getRawDataUnchecked();
+                        for (int i = 0, len = inRange.size(); i < len; i++) {
+                            final net.minecraft.server.level.ServerPlayer player = backingSet[i];
+                            if (player == null) continue;
+                            ++playerCap.computeIfAbsent(player, k -> new int[net.minecraft.server.level.ServerPlayer.MOBCATEGORY_TOTAL_ENUMS])[index];
+                        }
+                        */
+
+                        // Paper end - Optional per player mob spawns
+                    }
+                }
+            }
+        }
+
+        return new NaturalSpawner.SpawnState(spawnableChunkCount, map, potentialCalculator, null, chunkCap);
+    }
+    // Leaf end - optimize mob spawning
+
     static Biome getRoughBiome(BlockPos pos, ChunkAccess chunk) {
         return chunk.getNoiseBiome(QuartPos.fromBlock(pos.getX()), QuartPos.fromBlock(pos.getY()), QuartPos.fromBlock(pos.getZ())).value();
     }
@@ -149,7 +207,12 @@ public final class NaturalSpawner {
         return list;
     }
 
+    // Leaf start - optimize mob spawning - cache ChunkGenerator
     public static void spawnForChunk(ServerLevel level, LevelChunk chunk, NaturalSpawner.SpawnState spawnState, List<MobCategory> categories) {
+        spawnForChunk(level, chunk, spawnState, categories.toArray(net.minecraft.server.level.ServerChunkCache.EMPTY_MOB_CATEGORIES), level.chunkSource.getGenerator());
+    }
+    public static void spawnForChunk(ServerLevel level, LevelChunk chunk, NaturalSpawner.SpawnState spawnState, MobCategory[] categories, ChunkGenerator generator) {
+    // Leaf end - optimize mob spawning - cache ChunkGenerator
         for (MobCategory mobCategory : categories) {
             // Paper start - Optional per player mob spawns
             final boolean canSpawn;
@@ -192,8 +255,8 @@ public final class NaturalSpawner {
             }
             if (canSpawn) {
                 // Paper PR start - throttle failed spawn attempts
-                int spawnCount = spawnCategoryForChunk(mobCategory, level, chunk, spawnState::canSpawn, spawnState::afterSpawn,
-                    maxSpawns, level.paperConfig().entities.spawning.perPlayerMobSpawns ? level.getChunkSource().chunkMap::updatePlayerMobTypeMap : null, false);
+                int spawnCount = spawnCategoryForChunk(mobCategory, level, chunk, spawnState::canSpawn, spawnState::afterSpawn, // Leaf - optimize mob spawning - cache ChunkGenerator
+                    maxSpawns, level.paperConfig().entities.spawning.perPlayerMobSpawns ? level.getChunkSource().chunkMap::updatePlayerMobTypeMap : null, generator); // Leaf - optimize mob spawning - cache ChunkGenerator
                 if (spawnCount == 0) {
                     chunk.failedSpawnAttempts[mobCategory.ordinal()]++;
                 } else {
@@ -225,16 +288,16 @@ public final class NaturalSpawner {
         MobCategory category, ServerLevel level, LevelChunk chunk, NaturalSpawner.SpawnPredicate filter, NaturalSpawner.AfterSpawnCallback callback, final int maxSpawns, final Consumer<Entity> trackEntity
         // Paper PR start - throttle failed spawn attempts
     ) {
-        spawnCategoryForChunk(category, level, chunk, filter, callback, maxSpawns, trackEntity, false);
+        spawnCategoryForChunk(category, level, chunk, filter, callback, maxSpawns, trackEntity, level.chunkSource.getGenerator()); // Leaf - optimize mob spawning
     }
     public static int spawnCategoryForChunk(
-        MobCategory category, ServerLevel level, LevelChunk chunk, NaturalSpawner.SpawnPredicate filter, NaturalSpawner.AfterSpawnCallback callback, final int maxSpawns, final Consumer<Entity> trackEntity, final boolean nothing
+        MobCategory category, ServerLevel level, LevelChunk chunk, NaturalSpawner.SpawnPredicate filter, NaturalSpawner.AfterSpawnCallback callback, final int maxSpawns, final Consumer<Entity> trackEntity, final ChunkGenerator generator // Leaf - optimize mob spawning - cache ChunkGenerator
         // Paper PR end - throttle failed spawn attempts
     ) {
         // Paper end - Optional per player mob spawns
         BlockPos randomPosWithin = getRandomPosWithin(level, chunk);
         if (randomPosWithin.getY() >= level.getMinY() + 1) {
-            return spawnCategoryForPosition(category, level, chunk, randomPosWithin, filter, callback, maxSpawns, trackEntity, false); // Paper - Optional per player mob spawns // Paper PR - throttle failed spawn attempts
+            return spawnCategoryForPosition(category, level, chunk, randomPosWithin, filter, callback, maxSpawns, trackEntity, generator); // Paper - Optional per player mob spawns // Paper PR - throttle failed spawn attempts // Leaf - optimize mob spawning - cache ChunkGenerator
         }
 
         return 0; // Paper PR - throttle failed spawn attempts
@@ -259,55 +322,89 @@ public final class NaturalSpawner {
         MobCategory category, ServerLevel level, ChunkAccess chunk, BlockPos pos, NaturalSpawner.SpawnPredicate filter, NaturalSpawner.AfterSpawnCallback callback, final int maxSpawns, final @Nullable Consumer<Entity> trackEntity
         // Paper PR start - throttle failed spawn attempts
     ) {
-        spawnCategoryForPosition(category, level, chunk, pos, filter, callback, maxSpawns, trackEntity, false);
+        if (chunk instanceof LevelChunk levelChunk) { spawnCategoryForPosition(category, level, levelChunk, pos, filter, callback, maxSpawns, trackEntity, level.chunkSource.getGenerator()); } // Leaf - optimize mob spawning - cache ChunkGenerator
     }
     public static int spawnCategoryForPosition(
-        MobCategory category, ServerLevel level, ChunkAccess chunk, BlockPos pos, NaturalSpawner.SpawnPredicate filter, NaturalSpawner.AfterSpawnCallback callback, final int maxSpawns, final @Nullable Consumer<Entity> trackEntity, final boolean nothing
+        MobCategory category, ServerLevel level, LevelChunk chunk, BlockPos pos, NaturalSpawner.SpawnPredicate filter, NaturalSpawner.AfterSpawnCallback callback, final int maxSpawns, @Nullable final Consumer<Entity> trackEntity, final ChunkGenerator generator // Leaf - optimize mob spawning - cache ChunkGenerator
         // Paper PR end - throttle failed spawn attempts
     ) {
         // Paper end - Optional per player mob spawns
+        // Leaf start - optimize mob spawning
+        // ChunkGenerator generator = level.getChunkSource().getGenerator(); // cache ChunkGenerator
         StructureManager structureManager = level.structureManager();
-        ChunkGenerator generator = level.getChunkSource().getGenerator();
         int y = pos.getY();
+        int posX = pos.getX();
+        int posZ = pos.getZ();
         int i = 0; // Paper PR - throttle failed spawn attempts
-        BlockState blockState = level.getBlockStateIfLoadedAndInBounds(pos); // Paper - don't load chunks for mob spawn
+        BlockState blockState = ChunkPos.asLong(pos) == chunk.coordinateKey ? chunk.getBlockStateFinal(posX, y, posZ) : level.getBlockStateIfLoadedAndInBounds(pos); // Paper - don't load chunks for mob spawn
+        // Leaf end - optimize mob spawning
         if (blockState != null && !blockState.isRedstoneConductor(chunk, pos)) { // Paper - don't load chunks for mob spawn
             BlockPos.MutableBlockPos mutableBlockPos = new BlockPos.MutableBlockPos();
             //int i = 0; // Paper PR - throttle failed spawn attempts - move up
 
+            // Leaf start - optimize mob spawning
+            long rand = level.random.nextLong();
+            int bits = 0;
             for (int i1 = 0; i1 < 3; i1++) {
-                int x = pos.getX();
-                int z = pos.getZ();
-                int i2 = 6;
-                MobSpawnSettings.SpawnerData spawnerData = null;
-                SpawnGroupData spawnGroupData = null;
-                int ceil = Mth.ceil(level.random.nextFloat() * 4.0F);
+                //int i2 = 6;
+                //int ceil = Mth.ceil(level.random.nextFloat() * 4.0F);
+                int ceil = (int) (((rand >>> bits) & 0x3L) + 1);
+                bits += 2;
+                if (bits >= 62) {
+                    rand = level.random.nextLong();
+                    bits = 0;
+                }
                 int i3 = 0;
 
+                int x = posX;
+                int z = posZ;
+                MobSpawnSettings.SpawnerData spawnerData = null;
+                SpawnGroupData spawnGroupData = null;
                 for (int i4 = 0; i4 < ceil; i4++) {
-                    x += level.random.nextInt(6) - level.random.nextInt(6);
-                    z += level.random.nextInt(6) - level.random.nextInt(6);
-                    mutableBlockPos.set(x, y, z);
+                    int rand1=0,rand2=0,rand3=0,rand4=0,valuesNeeded=4;
+                    while (valuesNeeded > 0) {
+                        // [0, 61] 3 remains
+                        int threeBits = (int) ((rand >>> bits) & 0x7L);
+                        bits += 3;
+                        if (threeBits != 7 && threeBits != 6) {
+                            switch (valuesNeeded) {
+                                case 1 -> rand4 = threeBits;
+                                case 2 -> rand3 = threeBits;
+                                case 3 -> rand2 = threeBits;
+                                default -> rand1 = threeBits;
+                            }
+                            valuesNeeded--;
+                        }
+                        if (bits >= 62) {
+                            rand = level.random.nextLong();
+                            bits = 0;
+                        }
+                    }
+                    x += rand1 - rand2;
+                    z += rand3 - rand4;
+                    // x += level.random.nextInt(6) - level.random.nextInt(6);
+                    // z += level.random.nextInt(6) - level.random.nextInt(6);
                     double d = x + 0.5;
                     double d1 = z + 0.5;
-                    Player nearestPlayer = level.getNearestPlayer(d, y, d1, -1.0, level.purpurConfig.mobSpawningIgnoreCreativePlayers); // Purpur - mob spawning option to ignore creative players
-                    if (nearestPlayer != null) {
-                        double d2 = nearestPlayer.distanceToSqr(d, y, d1);
-                        if (level.isLoadedAndInBounds(mutableBlockPos) && isRightDistanceToPlayerAndSpawnPoint(level, chunk, mutableBlockPos, d2)) { // Paper - don't load chunks for mob spawn
+                    double d2 = level.natureSpawnChunkMap.nearest(d, y, d1); // Purpur - mob spawning option to ignore creative players
+                    if (d2 != Double.POSITIVE_INFINITY) {
+                        if (leaf$isRightDistanceToPlayerAndSpawnPoint(level, chunk, x, y, z, d2)) { // Paper - don't load chunks for mob spawn
+                            mutableBlockPos.set(x, y, z);
                             if (spawnerData == null) {
-                                Optional<MobSpawnSettings.SpawnerData> randomSpawnMobAt = getRandomSpawnMobAt(
-                                    level, structureManager, generator, category, level.random, mutableBlockPos
+                                MobSpawnSettings.SpawnerData randomSpawnMobAt = getRandomSpawnMobAt(
+                                    level, structureManager, generator, category, level.random, mutableBlockPos, chunk
                                 );
-                                if (randomSpawnMobAt.isEmpty()) {
+                                if (randomSpawnMobAt == null) {
                                     break;
                                 }
 
-                                spawnerData = randomSpawnMobAt.get();
+                                spawnerData = randomSpawnMobAt;
                                 ceil = spawnerData.minCount() + level.random.nextInt(1 + spawnerData.maxCount() - spawnerData.minCount());
                             }
+            // Leaf end - optimize mob spawning
 
                             // Paper start - PreCreatureSpawnEvent
-                            PreSpawnStatus doSpawning = isValidSpawnPostitionForType(level, category, structureManager, generator, spawnerData, mutableBlockPos, d2);
+                            PreSpawnStatus doSpawning = isValidSpawnPostitionForType(level, category, structureManager, generator, spawnerData, mutableBlockPos, d2, chunk); // Leaf - optimize mob spawning
                             // Paper start - per player mob count backoff
                             if (doSpawning == PreSpawnStatus.ABORT || doSpawning == PreSpawnStatus.CANCELLED) {
                                 level.getChunkSource().chunkMap.updateFailurePlayerMobTypeMap(mutableBlockPos.getX() >> 4, mutableBlockPos.getZ() >> 4, category);
@@ -372,6 +469,22 @@ public final class NaturalSpawner {
         }
     }
 
+    // Leaf start - optimize mob spawning
+    private static boolean leaf$isRightDistanceToPlayerAndSpawnPoint(ServerLevel level, ChunkAccess chunk, int x, int y, int z, double distance) {
+        if (distance <= 576.0 || distance == 16384.0) {
+            return false;
+        } else if (level.getSharedSpawnPos().distToCenterSqr(x + 0.5, y, z + 0.5) < 576.0) {
+            return false;
+        } else if (ChunkPos.asLong(x >> 4, z >> 4) == chunk.coordinateKey) {
+            return true;
+        } else if (level.getWorldBorder().isWithinBounds(new ChunkPos(x >> 4, z >> 4))) {
+            return level.chunkSource.getChunkAtIfLoadedImmediately(x >> 4, z >> 4).moonrise$getChunkHolder().isEntityTickingReady();
+        } else {
+            return false;
+        }
+    }
+    // Leaf end - optimize mob spawning
+
     // Paper start - PreCreatureSpawnEvent
     private enum PreSpawnStatus {
         FAIL,
@@ -389,6 +502,20 @@ public final class NaturalSpawner {
         BlockPos.MutableBlockPos pos,
         double distance
     ) {
+    // Leaf start - optimize mob spawning
+        return isValidSpawnPostitionForType(level, category, structureManager, generator, data, pos, distance, level.getChunk(pos.getX() >> 4, pos.getZ() >> 4));
+    }
+    private static PreSpawnStatus isValidSpawnPostitionForType(
+        ServerLevel level,
+        MobCategory category,
+        StructureManager structureManager,
+        ChunkGenerator generator,
+        MobSpawnSettings.SpawnerData data,
+        BlockPos.MutableBlockPos pos,
+        double distance,
+        LevelChunk chunk
+    ) {
+    // Leaf end - optimize mob spawning
         EntityType<?> entityType = data.type();
         // Paper start - PreCreatureSpawnEvent
         com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent event = new com.destroystokyo.paper.event.entity.PreCreatureSpawnEvent(
@@ -408,7 +535,7 @@ public final class NaturalSpawner {
                     || !(distance > entityType.getCategory().getDespawnDistance() * entityType.getCategory().getDespawnDistance())
             )
             && entityType.canSummon()
-            && canSpawnMobAt(level, structureManager, generator, category, data, pos)
+            && mobsAt(level, structureManager, generator, category, pos, null, chunk).contains(data) // Leaf - optimize mob spawning
             && SpawnPlacements.isSpawnPositionOk(entityType, level, pos)
             && SpawnPlacements.checkSpawnRules(entityType, level, EntitySpawnReason.NATURAL, pos, level.random)
             && level.noCollision(entityType.getSpawnAABB(pos.getX() + 0.5, pos.getY(), pos.getZ() + 0.5));
@@ -449,6 +576,17 @@ public final class NaturalSpawner {
             : mobsAt(level, structureManager, generator, category, pos, biome).getRandom(random);
     }
 
+    // Leaf start - optimize mob spawning
+    private static MobSpawnSettings.SpawnerData getRandomSpawnMobAt(
+        ServerLevel level, StructureManager structureManager, ChunkGenerator generator, MobCategory category, RandomSource random, BlockPos pos, LevelChunk chunk
+    ) {
+        Holder<Biome> biome = org.dreeam.leaf.config.modules.opt.OptimizeBiome.mobSpawn ? level.getBiomeCached(chunk, pos) : level.getBiome(pos); // Leaf - cache getBiome
+        return category == MobCategory.WATER_AMBIENT && biome.is(BiomeTags.REDUCED_WATER_AMBIENT_SPAWNS) && random.nextFloat() < 0.98F
+            ? null
+            : mobsAt(level, structureManager, generator, category, pos, biome, chunk).getRandomNullable(random);
+    }
+    // Leaf end - optimize mob spawning
+
     private static boolean canSpawnMobAt(
         ServerLevel level, StructureManager structureManager, ChunkGenerator generator, MobCategory category, MobSpawnSettings.SpawnerData data, BlockPos pos
     ) {
@@ -463,8 +601,22 @@ public final class NaturalSpawner {
             : generator.getMobsAt(biome != null ? biome : (org.dreeam.leaf.config.modules.opt.OptimizeBiome.mobSpawn ? level.getBiomeCached(null, pos) : level.getBiome(pos)), structureManager, cetagory, pos); // Leaf - cache getBiome
     }
 
+    // Leaf start - optimize mob spawning
+    private static WeightedList<MobSpawnSettings.SpawnerData> mobsAt(
+        ServerLevel level, StructureManager structureManager, ChunkGenerator generator, MobCategory cetagory, BlockPos pos, @Nullable Holder<Biome> biome, LevelChunk chunk
+    ) {
+        return isInNetherFortressBounds(pos, level, cetagory, structureManager, chunk)
+            ? NetherFortressStructure.FORTRESS_ENEMIES
+            : generator.getMobsAtChunk(biome != null ? biome : (org.dreeam.leaf.config.modules.opt.OptimizeBiome.mobSpawn ? level.getBiomeCached(chunk, pos) : level.getBiome(pos)), structureManager, cetagory, pos, chunk); // Leaf - cache getBiome
+    }
+    // Leaf end - optimize mob spawning
+
     public static boolean isInNetherFortressBounds(BlockPos pos, ServerLevel level, MobCategory category, StructureManager structureManager) {
-        if (category == MobCategory.MONSTER && level.getBlockState(pos.below()).is(Blocks.NETHER_BRICKS)) {
+        // Leaf start - optimize mob spawning
+        if (category == MobCategory.MONSTER) {
+            BlockState blockState = level.getBlockStateIfLoaded(pos.below());
+            if (blockState == null || !blockState.is(Blocks.NETHER_BRICKS)) return false;
+        // Leaf end - optimize mob spawning
             Structure structure = structureManager.registryAccess().lookupOrThrow(Registries.STRUCTURE).getValue(BuiltinStructures.FORTRESS);
             return structure != null && structureManager.getStructureAt(pos, structure).isValid();
         } else {
@@ -472,6 +624,19 @@ public final class NaturalSpawner {
         }
     }
 
+    // Leaf start - optimize mob spawning
+    public static boolean isInNetherFortressBounds(BlockPos pos, ServerLevel level, MobCategory category, StructureManager structureManager, LevelChunk chunk) {
+        if (category == MobCategory.MONSTER) {
+            BlockState blockState = chunk.coordinateKey == ChunkPos.asLong(pos) ? chunk.getBlockStateFinal(pos.getX(), pos.getY() - 1, pos.getZ()) : level.getBlockStateIfLoaded(pos.below());
+            if (blockState == null || !blockState.is(Blocks.NETHER_BRICKS)) return false;
+            Structure structure = structureManager.registryAccess().lookupOrThrow(Registries.STRUCTURE).getValue(BuiltinStructures.FORTRESS);
+            return structure != null && structureManager.getStructureAt(pos, structure).isValid();
+        } else {
+            return false;
+        }
+    }
+    // Leaf end - optimize mob spawning
+
     private static BlockPos getRandomPosWithin(Level level, LevelChunk chunk) {
         ChunkPos pos = chunk.getPos();
         int i = pos.getMinBlockX() + level.random.nextInt(16);
@@ -612,6 +777,7 @@ public final class NaturalSpawner {
         @Nullable
         private EntityType<?> lastCheckedType;
         private double lastCharge;
+        private final it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap<int[]> chunkCap; // Leaf - optimize mob spawning
 
         SpawnState(
             int spawnableChunkCount,
@@ -624,8 +790,26 @@ public final class NaturalSpawner {
             this.spawnPotential = spawnPotential;
             this.localMobCapCalculator = localMobCapCalculator;
             this.unmodifiableMobCategoryCounts = Object2IntMaps.unmodifiable(mobCategoryCounts);
+            this.chunkCap = new it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap<>(); // Leaf - optimize mob spawning
         }
 
+        // Leaf start - optimize mob spawning
+        SpawnState(
+            int spawnableChunkCount,
+            Object2IntOpenHashMap<MobCategory> mobCategoryCounts,
+            PotentialCalculator spawnPotential,
+            LocalMobCapCalculator localMobCapCalculator,
+            it.unimi.dsi.fastutil.longs.Long2ObjectOpenHashMap<int[]> playerCap
+        ) {
+            this.spawnableChunkCount = spawnableChunkCount;
+            this.mobCategoryCounts = mobCategoryCounts;
+            this.spawnPotential = spawnPotential;
+            this.localMobCapCalculator = localMobCapCalculator;
+            this.unmodifiableMobCategoryCounts = Object2IntMaps.unmodifiable(mobCategoryCounts);
+            this.chunkCap = playerCap; // Leaf - optimize mob spawning
+        }
+        // Leaf end - optimize mob spawning
+
         private boolean canSpawn(EntityType<?> entityType, BlockPos pos, ChunkAccess chunk) {
             this.lastCheckedPos = pos;
             this.lastCheckedType = entityType;
@@ -680,5 +864,32 @@ public final class NaturalSpawner {
         boolean canSpawnForCategoryLocal(MobCategory category, ChunkPos chunkPos) {
             return this.localMobCapCalculator.canSpawn(category, chunkPos);
         }
+
+        // Leaf start - optimize mob spawning
+        public void applyPerPlayerMobCount(ServerLevel level) {
+            if (chunkCap.isEmpty()) {
+                return;
+            }
+            final var iterator = chunkCap.long2ObjectEntrySet().fastIterator();
+            final ca.spottedleaf.moonrise.common.misc.NearbyPlayers nearbyPlayers = level.moonrise$getNearbyPlayers();
+            while (iterator.hasNext()) {
+                var entry = iterator.next();
+                long chunk = entry.getLongKey();
+                int[] cap = entry.getValue();
+                ca.spottedleaf.moonrise.common.list.ReferenceList<net.minecraft.server.level.ServerPlayer> players = nearbyPlayers.getPlayersByChunk(ChunkPos.getX(chunk), ChunkPos.getZ(chunk), ca.spottedleaf.moonrise.common.misc.NearbyPlayers.NearbyMapType.TICK_VIEW_DISTANCE);
+                if (players == null) {
+                    continue;
+                }
+                int playersSize = players.size();
+                net.minecraft.server.level.ServerPlayer[] playersRawDataUnchecked = players.getRawDataUnchecked();
+                for (int i = 0; i < playersSize; i++) {
+                    int[] p = playersRawDataUnchecked[i].mobCounts;
+                    for (int j = 0; j < net.minecraft.server.level.ServerPlayer.MOBCATEGORY_TOTAL_ENUMS; j++) {
+                        p[j] += cap[j];
+                    }
+                }
+            }
+        }
+        // Leaf end - optimize mob spawning
     }
 }
diff --git a/net/minecraft/world/level/StructureManager.java b/net/minecraft/world/level/StructureManager.java
index fbe93098ce0366054a6da857cd808af1431b6612..57de70773d2766a8f3a41e61efc16ceb7a9f80c8 100644
--- a/net/minecraft/world/level/StructureManager.java
+++ b/net/minecraft/world/level/StructureManager.java
@@ -90,6 +90,7 @@ public class StructureManager {
 
     @Nullable
     public StructureStart getStartForStructure(SectionPos sectionPos, Structure structure, StructureAccess structureAccess) {
+        // Leaf - optimize mob spawning - diff
         return structureAccess.getStartForStructure(structure);
     }
 
@@ -181,6 +182,12 @@ public class StructureManager {
         //SectionPos sectionPos = SectionPos.of(pos); // Leaf - optimise ChunkGenerator#getMobsAt
         return this.level.getChunk(pos.getX() >> 4, pos.getZ() >> 4, ChunkStatus.STRUCTURE_REFERENCES).getAllReferences(); // Leaf - optimise ChunkGenerator#getMobsAt
     }
+    // Leaf start - optimize mob spawning
+    public Map<Structure, LongSet> getAllStructuresAtChunk(net.minecraft.world.level.chunk.ChunkAccess chunk) {
+        //SectionPos sectionPos = SectionPos.of(pos); // Leaf - optimise ChunkGenerator#getMobsAt
+        return chunk.getAllReferences(); // Leaf - optimise ChunkGenerator#getMobsAt
+    }
+    // Leaf end - optimize mob spawning
 
     public StructureCheckResult checkStructurePresence(ChunkPos chunkPos, Structure structure, StructurePlacement placement, boolean skipKnownStructures) {
         return this.structureCheck.checkStart(chunkPos, structure, placement, skipKnownStructures);
diff --git a/net/minecraft/world/level/biome/MobSpawnSettings.java b/net/minecraft/world/level/biome/MobSpawnSettings.java
index db3b8a237d63255aa9ffd70c88a093002a6bd770..4a69f404eee00d8972e9501a76031d4339136b6f 100644
--- a/net/minecraft/world/level/biome/MobSpawnSettings.java
+++ b/net/minecraft/world/level/biome/MobSpawnSettings.java
@@ -52,7 +52,7 @@ public class MobSpawnSettings {
         Map<EntityType<?>, MobSpawnSettings.MobSpawnCost> mobSpawnCosts
     ) {
         this.creatureGenerationProbability = creatureGenerationProbability;
-        this.spawners = ImmutableMap.copyOf(spawners);
+        this.spawners = com.google.common.collect.Maps.immutableEnumMap(spawners); // Leaf - optimize mob spawning
         this.mobSpawnCosts = ImmutableMap.copyOf(mobSpawnCosts);
     }
 
diff --git a/net/minecraft/world/level/chunk/ChunkAccess.java b/net/minecraft/world/level/chunk/ChunkAccess.java
index 7fc5e56586f2fe8ea1a7437b42da893383878880..529cbf81384c4b1da2366fea26f66d94d9a71118 100644
--- a/net/minecraft/world/level/chunk/ChunkAccess.java
+++ b/net/minecraft/world/level/chunk/ChunkAccess.java
@@ -92,7 +92,7 @@ public abstract class ChunkAccess implements BiomeManager.NoiseBiomeSource, Ligh
     private boolean slimeChunk;
     private boolean hasComputedSlimeChunk;
     // Leaf end - Matter - Secure Seed
-    public final long[] failedSpawnAttempts = new long[net.minecraft.world.entity.MobCategory.values().length]; // Paper PR - throttle failed spawn attempts
+    public final long[] failedSpawnAttempts = new long[net.minecraft.server.level.ServerPlayer.MOBCATEGORY_TOTAL_ENUMS]; // Paper PR - throttle failed spawn attempts // Leaf - optimize mob spawning
 
     // Paper start - rewrite chunk system
     private volatile ca.spottedleaf.moonrise.patches.starlight.light.SWMRNibbleArray[] blockNibbles;
diff --git a/net/minecraft/world/level/chunk/ChunkGenerator.java b/net/minecraft/world/level/chunk/ChunkGenerator.java
index fb77cd58542c1a690cbeaa6ed3a4d657de6e619d..bd0278c9c1466291eb65d37afc90096f5abcbd56 100644
--- a/net/minecraft/world/level/chunk/ChunkGenerator.java
+++ b/net/minecraft/world/level/chunk/ChunkGenerator.java
@@ -516,6 +516,35 @@ public abstract class ChunkGenerator {
         return biome.value().getMobSettings().getMobs(category);
     }
 
+    // Leaf start - optimize mob spawning
+    public WeightedList<MobSpawnSettings.SpawnerData> getMobsAtChunk(Holder<Biome> biome, StructureManager structureManager, MobCategory category, BlockPos pos, ChunkAccess chunk) {
+        Map<Structure, LongSet> allStructuresAt = ChunkPos.asLong(pos) == chunk.coordinateKey ? structureManager.getAllStructuresAtChunk(chunk) : structureManager.getAllStructuresAt(pos);
+
+        if (!allStructuresAt.isEmpty()) for (Entry<Structure, LongSet> entry : allStructuresAt.entrySet()) {
+            Structure structure = entry.getKey();
+            StructureSpawnOverride structureSpawnOverride = structure.spawnOverrides().get(category);
+            if (structureSpawnOverride != null) {
+                // Leaf start - optimise ChunkGenerator#getMobsAt
+                for (long l : entry.getValue()) {
+                    StructureStart startForStructure = structureManager.getStartForStructure(
+                        null, structure, chunk.getPos().longKey == l ? chunk : structureManager.level.getChunk(ChunkPos.getX(l), ChunkPos.getZ(l), ChunkStatus.STRUCTURE_STARTS)
+                    );
+                    if (startForStructure != null && startForStructure.isValid()) {
+                        if (structureSpawnOverride.boundingBox() == StructureSpawnOverride.BoundingBoxType.PIECE
+                            ? structureManager.structureHasPieceAt(pos, startForStructure)
+                            : startForStructure.getBoundingBox().isInside(pos)) {
+                            return structureSpawnOverride.spawns();
+                        }
+                    }
+                }
+                // Leaf end - optimise ChunkGenerator#getMobsAt
+            }
+        }
+
+        return biome.value().getMobSettings().getMobs(category);
+    }
+    // Leaf end - optimize mob spawning
+
     public void createStructures(
         RegistryAccess registryAccess,
         ChunkGeneratorStructureState structureState,
diff --git a/net/minecraft/world/level/entity/EntityTickList.java b/net/minecraft/world/level/entity/EntityTickList.java
index 5a90b3bffeeb08a168b370e49d18c5f8b257a980..ba173bc1751c495e6fa497566b5ed3c7a9547364 100644
--- a/net/minecraft/world/level/entity/EntityTickList.java
+++ b/net/minecraft/world/level/entity/EntityTickList.java
@@ -9,7 +9,7 @@ import javax.annotation.Nullable;
 import net.minecraft.world.entity.Entity;
 
 public class EntityTickList {
-    public final ca.spottedleaf.moonrise.common.list.IteratorSafeOrderedReferenceSet<net.minecraft.world.entity.Entity> entities = new ca.spottedleaf.moonrise.common.list.IteratorSafeOrderedReferenceSet<>(org.dreeam.leaf.config.modules.async.AsyncMobSpawning.enabled); // Paper - rewrite chunk system // Pufferfish - private->public and do thread check
+    private final ca.spottedleaf.moonrise.common.list.IteratorSafeOrderedReferenceSet<net.minecraft.world.entity.Entity> entities = new ca.spottedleaf.moonrise.common.list.IteratorSafeOrderedReferenceSet<>(); // Paper - rewrite chunk system
 
     // Leaf start - SparklyPaper - parallel world ticking mod
     // preserve original constructor
